src/oauth2c_token_cache_gen_server.erl

```diff
@@ -0,0 +1,144 @@
+%%% @end
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+%%%_* Module declaration ===============================================
+-module(oauth2c_token_cache_gen_server).
+-behaviour(gen_server).
+
+%%%_* Exports ==========================================================
+%%%_ * API -------------------------------------------------------------
+-export([start/0]).
+-export([start/1]).
+-export([stop/0]).
+-export([get/1]).
+-export([insert/2]).
+-export([delete/1]).
+
+%% gen_server
+-export([init/1]).
+-export([terminate/2]).
+-export([handle_call/3]).
+-export([handle_cast/2]).
+-export([handle_info/2]).
+-export([code_change/3]).
+
+%%%_* Includes =========================================================
+%%%_* Code =============================================================
+%%%_ * Types -----------------------------------------------------------
+%%%_ * API -------------------------------------------------------------
+
+start() -> start(#{cache_ttl => 3600000, cache => #{}}).
+start(State) -> gen_server:start_link({local, ?MODULE}, ?MODULE, State, []).
+stop() -> gen_server:cast(?MODULE, stop).
+get(Key) -> gen_server:call(?MODULE, {get, Key}).
+insert(Key, Value) -> gen_server:cast(?MODULE, {insert, {Key, Value}}).
+delete(Key) -> gen_server:cast(?MODULE, {delete, Key}).
+
+%%%_ * gen_server callbacks --------------------------------------------
+
+init(State) ->
+  {ok, State}.
+
+handle_call({get, Key}, _From, State = #{cache := Cache, cache_ttl := TTL}) ->
+  case get_token(Cache, Key, TTL, os:system_time(millisecond)) of
+    {ok, Token} ->
+      {reply, {ok, Token}, State};
+    token_expired ->
+      {reply, not_found, State#{cache := maps:remove(Key, Cache)}};
+    not_found ->
+      {reply, not_found, State}
+  end.
+
+handle_cast({insert, {Key, Value}}, State = #{cache := Cache}) ->
+  NewCache = update_cache(Cache, Key, Value),
+  {noreply, State#{cache := NewCache}};
+
+handle_cast({delete, Key}, State = #{cache := Cache}) ->
+  {noreply, State#{cache := maps:without([Key], Cache)}};
+
+handle_cast(stop, State) ->
+  {stop, normal, ok, State};
+
+handle_cast(_Msg, State) ->
+  {noreply, State}.
+
+handle_info(_Info, State) ->
+  {noreply, State}.
+
+terminate(_Reason, _State) ->
+  ok.
+
+code_change(_OldVsn, State, _Extra) ->
+  {ok, State}.
+
+%%%_ * Private functions -----------------------------------------------
+
+-spec update_cache(map(), atom(), any()) -> map().
+update_cache(Cache, Key, Value) ->
+  Cache#{Key => {Value, os:system_time(millisecond)}}.
+
+-spec get_token(map(), atom(), integer(), integer())
+  -> atom() | {atom(), any()}.
+get_token(Cache, Key, TTL, Now) ->
+  case maps:is_key(Key, Cache) of
+    true -> get_token_if_not_expired(maps:get(Key, Cache), TTL, Now);
+    false -> not_found
+  end.
+
+-spec get_token_if_not_expired({any(), integer()}, integer(), integer())
+  -> atom() | {ok, atom()}.
+% Token exist and is valid.
```

Comment:
The funcion header itself (function name and `when` clause) makes it clear enough, so the function clause comments aren't necessary. If you really want to have them, I would move them to the beginning of the body of each clause instead, since it's easy to take this comment as a comment to the function as a whole rather than the clause.

Reply:
I have removed the comments.